File: src/Devscribed.Admin.Web/Services/VacationService.cs
using Devscribed.Admin.Web.Data;
using Devscribed.Admin.Web.Models;
using Devscribed.Admin.Web.Validation;
using Microsoft.EntityFrameworkCore;

namespace Devscribed.Admin.Web.Services;

public class VacationService
{
    private readonly AppDbContext _db;

    public VacationService(AppDbContext db)
    {
        _db = db;
    }

    public async Task<VacationFetchResult> GetVacationAsync(
        Guid organizationId, Guid targetMembershipId, Guid callerMembershipId, string callerRole)
    {
        var target = await _db.Memberships
            .SingleOrDefaultAsync(m => m.Id == targetMembershipId && m.OrganizationId == organizationId);

        if (target == null)
            return VacationFetchResult.NotFound();

        var isOwnMembership = target.Id == callerMembershipId;
        if (!MemberPermissions.CanViewVacation(callerRole, isOwnMembership))
            return VacationFetchResult.Forbidden("You do not have permission to view this member's vacation data");

        var isFullView = callerRole is "admin" or "manager";
        var financials = await _db.MemberFinancials.SingleOrDefaultAsync(f => f.MembershipId == targetMembershipId);

        var dto = new VacationDto
        {
            Financials = (financials != null && isFullView)
                ? new VacationFinancialsDto
                {
                    MonthlySalary = financials.MonthlySalary,
                    ClientHourlyRate = financials.ClientHourlyRate,
                    VacationReservePercent = financials.VacationReservePercent,
                    IsReservePercentManual = financials.IsReservePercentManual,
                    VacationDaysPerYear = financials.VacationDaysPerYear,
                    Currency = financials.Currency,
                }
                : null,
            Balance = financials == null
                ? null
                : new VacationBalanceDto
                {
                    ReserveBalance = isFullView ? 0m : null,
                    AvailableDays = 0,
                    UsedDays = 0,
                    PendingDays = 0,
                    TotalDaysPerYear = financials.VacationDaysPerYear,
                },
            CanEdit = MemberPermissions.CanEditMemberFinancials(callerRole),
            CanReviewRequests = false,
            CanSubmitRequest = false,
        };

        return VacationFetchResult.Ok(dto);
    }

    public async Task<MemberFinancialsActionResult> UpdateFinancialsAsync(
        Guid organizationId, Guid targetMembershipId, Guid callerAccountId, string callerRole, UpdateMemberFinancialsRequest request)
    {
        if (!MemberPermissions.CanEditMemberFinancials(callerRole))
            return MemberFinancialsActionResult.Forbidden("You do not have permission to edit financial settings");

        var errors = MemberFinancialsValidator.ValidateAll(request);
        if (errors.Count > 0)
            return MemberFinancialsActionResult.FieldValidationError(errors);

        var target = await _db.Memberships
            .SingleOrDefaultAsync(m => m.Id == targetMembershipId && m.OrganizationId == organizationId);

        if (target == null)
            return MemberFinancialsActionResult.NotFound();

        if (target.Status == "removed")
            return MemberFinancialsActionResult.BadRequest("member_removed", "Cannot configure vacation for a removed member");

        var financials = await _db.MemberFinancials.SingleOrDefaultAsync(f => f.MembershipId == targetMembershipId);
        if (financials == null)
        {
            financials = new MemberFinancials { Id = Guid.NewGuid(), MembershipId = targetMembershipId };
            _db.MemberFinancials.Add(financials);
        }

        var effectivePercent = request.IsReservePercentManual
            ? request.VacationReservePercent!.Value
            : VacationReserveCalculator.CalculateReservePercent(
                request.MonthlySalary!.Value, request.ClientHourlyRate!.Value, request.VacationDaysPerYear!.Value);

        financials.MonthlySalary = request.MonthlySalary!.Value;
        financials.ClientHourlyRate = request.ClientHourlyRate!.Value;
        financials.VacationDaysPerYear = request.VacationDaysPerYear!.Value;
        financials.Currency = request.Currency!;
        financials.IsReservePercentManual = request.IsReservePercentManual;
        financials.VacationReservePercent = effectivePercent;
        financials.UpdatedAt = DateTime.UtcNow;
        financials.UpdatedByAccountId = callerAccountId;

        _db.MemberFinancialsSnapshots.Add(new MemberFinancialsSnapshot
        {
            Id = Guid.NewGuid(),
            MembershipId = targetMembershipId,
            MonthlySalary = financials.MonthlySalary,
            ClientHourlyRate = financials.ClientHourlyRate,
            VacationReservePercent = financials.VacationReservePercent,
            IsReservePercentManual = financials.IsReservePercentManual,
            VacationDaysPerYear = financials.VacationDaysPerYear,
            Currency = financials.Currency,
            EffectiveFrom = DateOnly.FromDateTime(DateTime.UtcNow),
            CreatedAt = DateTime.UtcNow,
        });

        await _db.SaveChangesAsync();

        return MemberFinancialsActionResult.Ok(effectivePercent);
    }
}

public class VacationFinancialsDto
{
    public decimal MonthlySalary { get; set; }
    public decimal ClientHourlyRate { get; set; }
    public decimal VacationReservePercent { get; set; }
    public bool IsReservePercentManual { get; set; }
    public int VacationDaysPerYear { get; set; }
    public string Currency { get; set; } = string.Empty;
}

public class VacationBalanceDto
{
    public decimal? ReserveBalance { get; set; }
    public int AvailableDays { get; set; }
    public int UsedDays { get; set; }
    public int PendingDays { get; set; }
    public int TotalDaysPerYear { get; set; }
}

public class VacationDto
{
    public VacationFinancialsDto? Financials { get; set; }
    public VacationBalanceDto? Balance { get; set; }
    public bool CanEdit { get; set; }
    public bool CanReviewRequests { get; set; }
    public bool CanSubmitRequest { get; set; }
}

public enum VacationFetchOutcome
{
    Success,
    NotFound,
    Forbidden,
}

public class VacationFetchResult
{
    public VacationFetchOutcome Outcome { get; init; }
    public VacationDto? Dto { get; init; }
    public string? ErrorMessage { get; init; }

    public static VacationFetchResult Ok(VacationDto dto) => new() { Outcome = VacationFetchOutcome.Success, Dto = dto };
    public static VacationFetchResult NotFound() => new() { Outcome = VacationFetchOutcome.NotFound };
    public static VacationFetchResult Forbidden(string message) => new()
    { Outcome = VacationFetchOutcome.Forbidden, ErrorMessage = message };
}

public enum MemberFinancialsActionOutcome
{
    Success,
    NotFound,
    Forbidden,
    BadRequest,
    FieldValidation,
}

public class MemberFinancialsActionResult
{
    public MemberFinancialsActionOutcome Outcome { get; init; }
    public string? ErrorCode { get; init; }
    public string? ErrorMessage { get; init; }
    public Dictionary<string, string>? FieldErrors { get; init; }
    public decimal VacationReservePercent { get; init; }

    public static MemberFinancialsActionResult Ok(decimal vacationReservePercent) => new()
    { Outcome = MemberFinancialsActionOutcome.Success, VacationReservePercent = vacationReservePercent };
    public static MemberFinancialsActionResult NotFound() => new() { Outcome = MemberFinancialsActionOutcome.NotFound };
    public static MemberFinancialsActionResult Forbidden(string message, string code = "forbidden") => new()
    { Outcome = MemberFinancialsActionOutcome.Forbidden, ErrorCode = code, ErrorMessage = message };
    public static MemberFinancialsActionResult BadRequest(string code, string message) => new()
    { Outcome = MemberFinancialsActionOutcome.BadRequest, ErrorCode = code, ErrorMessage = message };
    public static MemberFinancialsActionResult FieldValidationError(Dictionary<string, string> errors) => new()
    { Outcome = MemberFinancialsActionOutcome.FieldValidation, FieldErrors = errors };
}

File: apps/api/src/vacation/vacation.service.ts
import {
  BadRequestException,
  ForbiddenException,
  Injectable,
  NotFoundException,
} from '@nestjs/common';
import {
  FINANCIALS_MESSAGES,
  MEMBER_MESSAGES,
  calculateReservePercent,
  can,
  validateMemberFinancials,
  type MemberFinancialsInput,
  type Role,
} from '@devscribed/validation';
import type { MemberFinancials } from '@prisma/client';
import type { SessionPayload } from '../auth/session.service';
import { PrismaService } from '../prisma.service';

/** Default `VacationDaysPerYear` (spec 07 requirement 4) — the Prisma column default. */
const DEFAULT_VACATION_DAYS_PER_YEAR = 20;

/** PUT `.../vacation/financials` request body — every field is optional at the wire level. */
export type MemberFinancialsDto = Partial<MemberFinancialsInput>;

/** Full financial block returned to admin/manager (numbers, never Prisma `Decimal`). */
interface FinancialsBlock {
  monthlySalary: number;
  clientHourlyRate: number;
  vacationReservePercent: number;
  isReservePercentManual: boolean;
  vacationDaysPerYear: number;
  currency: string;
}

/** Balance block — every accrual figure is zero until spec 08 lands. */
interface BalanceBlock {
  reserveBalance: number | null;
  availableDays: number;
  usedDays: number;
  pendingDays: number;
  totalDaysPerYear: number;
}

export interface VacationView {
  financials: FinancialsBlock | null;
  balance: BalanceBlock | null;
  canEdit: boolean;
  canReviewRequests: boolean;
  canSubmitRequest: boolean;
}

interface CallerMembership {
  id: string;
  role: Role;
  organizationId: string;
  accountId: string;
}

/**
 * Spec 07 — Member Financial Settings. Reads and writes the Vacation tab's financial
 * data. Mirrors `MembersService`: caller resolved from the session (never the URL), the
 * org-row `SELECT ... FOR UPDATE` lock on writes, and the `{ error, message }` /
 * `{ errors }` exception shapes.
 */
@Injectable()
export class VacationService {
  constructor(private readonly prisma: PrismaService) {}

  /**
   * `GET /members/:memberId/vacation`. admin/manager see full financials; a `user` sees
   * only their own balance (days, no money); `viewer` — and a `user` viewing another
   * member — get 403 (TC-07-INT-07/08).
   */
  async getVacation(session: SessionPayload, targetId: string): Promise<VacationView> {
    const caller = await this.requireCaller(session);

    const target = await this.prisma.membership.findFirst({
      where: { id: targetId, organizationId: caller.organizationId },
    });
    if (!target) {
      throw new NotFoundException({ error: 'not_found', message: MEMBER_MESSAGES.memberNotFound });
    }

    const canViewAny = can(caller.role, 'view-vacation');
    const isSelf = target.id === caller.id;
    // viewer: no own-balance capability either → forbidden. user: own-balance only.
    if (!canViewAny && !(isSelf && can(caller.role, 'view-own-vacation-balance'))) {
      throw new ForbiddenException({
        error: 'forbidden',
        message: FINANCIALS_MESSAGES.viewForbidden,
      });
    }

    const financials = await this.prisma.memberFinancials.findUnique({
      where: { membershipId: target.id },
    });

    const targetActive = target.status === 'active';

    if (canViewAny) {
      // admin/manager — full financial view.
      const canEdit = can(caller.role, 'edit-member-financials') && targetActive;
      if (!financials) {
        return {
          financials: null,
          balance: null,
          canEdit,
          canReviewRequests: false,
          canSubmitRequest: false,
        };
      }
      return {
        financials: this.toFinancialsBlock(financials),
        balance: this.zeroBalance(financials.vacationDaysPerYear, false),
        canEdit,
        canReviewRequests: false,
        canSubmitRequest: false,
      };
    }

    // user viewing own membership — days only, never money.
    return {
      financials: null,
      balance: financials ? this.zeroBalance(financials.vacationDaysPerYear, true) : null,
      canEdit: false,
      canReviewRequests: false,
      canSubmitRequest: false,
    };
  }

  /**
   * `PUT /members/:memberId/vacation/financials`. Upserts the settings and appends an
   * immutable snapshot with `EffectiveFrom` = today. In auto mode the reserve percent is
   * recomputed from salary/rate/days; in manual mode the validated value is stored.
   */
  async updateFinancials(
    session: SessionPayload,
    targetId: string,
    dto: MemberFinancialsDto,
  ): Promise<{ success: true; vacationReservePercent: number }> {
    const caller = await this.requireCaller(session);
    if (!can(caller.role, 'edit-member-financials')) {
      throw new ForbiddenException({
        error: 'forbidden',
        message: FINANCIALS_MESSAGES.editForbidden,
      });
    }

    return this.prisma.$transaction(async (tx) => {
      // Same org-row lock the members flows take — serializes concurrent writes against
      // this membership's single financials/snapshot chain.
      await tx.$queryRaw`SELECT id FROM "Organization" WHERE id = ${caller.organizationId} FOR UPDATE`;

      const target = await tx.membership.findFirst({
        where: { id: targetId, organizationId: caller.organizationId },
      });
      if (!target) {
        throw new NotFoundException({ error: 'not_found', message: MEMBER_MESSAGES.memberNotFound });
      }
      if (target.status === 'removed') {
        throw new BadRequestException({
          error: 'member_removed',
          message: FINANCIALS_MESSAGES.memberRemoved,
        });
      }

      const validation = validateMemberFinancials(dto);
      if (!validation.valid) {
        throw new BadRequestException({ errors: validation.errors });
      }

      const { monthlySalary, clientHourlyRate, vacationDaysPerYear, currency, isReservePercentManual } =
        validation.value;

      // Auto mode recomputes from the freshly validated inputs (requirements 7-8); manual
      // mode stores the value the manager supplied (requirement 9) — already validated non-null.
      const vacationReservePercent = isReservePercentManual
        ? (validation.value.vacationReservePercent as number)
        : calculateReservePercent({ monthlySalary, clientHourlyRate, vacationDaysPerYear });

      const values = {
        monthlySalary,
        clientHourlyRate,
        vacationReservePercent,
        isReservePercentManual,
        vacationDaysPerYear,
        currency,
      };

      await tx.memberFinancials.upsert({
        where: { membershipId: target.id },
        create: { membershipId: target.id, ...values, updatedByAccountId: caller.accountId },
        update: { ...values, updatedByAccountId: caller.accountId },
      });

      // Every write appends an immutable snapshot (requirement 11); prior snapshots are
      // never touched. `effectiveFrom` is today, date-only (midnight UTC).
      await tx.memberFinancialsSnapshot.create({
        data: { membershipId: target.id, ...values, effectiveFrom: this.today() },
      });

      return { success: true as const, vacationReservePercent };
    });
  }

  /** Prisma `Decimal` → `number` for the wire. */
  private toFinancialsBlock(f: MemberFinancials): FinancialsBlock {
    return {
      monthlySalary: f.monthlySalary.toNumber(),
      clientHourlyRate: f.clientHourlyRate.toNumber(),
      vacationReservePercent: f.vacationReservePercent.toNumber(),
      isReservePercentManual: f.isReservePercentManual,
      vacationDaysPerYear: f.vacationDaysPerYear,
      currency: f.currency,
    };
  }

  /** All accrual figures are zero pre-spec-08; `user` never sees the monetary reserve. */
  private zeroBalance(totalDaysPerYear: number, hideReserve: boolean): BalanceBlock {
    return {
      reserveBalance: hideReserve ? null : 0,
      availableDays: 0,
      usedDays: 0,
      pendingDays: 0,
      totalDaysPerYear: totalDaysPerYear ?? DEFAULT_VACATION_DAYS_PER_YEAR,
    };
  }

  /** Today as a date-only value (midnight UTC), for the snapshot's `@db.Date` column. */
  private today(): Date {
    const now = new Date();
    return new Date(Date.UTC(now.getUTCFullYear(), now.getUTCMonth(), now.getUTCDate()));
  }

  /** Caller's own membership, resolved from the session — mirrors `MembersService`. */
  private async requireCaller(session: SessionPayload): Promise<CallerMembership> {
    const caller = await this.prisma.membership.findUnique({
      where: { accountId: session.accountId },
    });
    if (!caller || caller.status !== 'active' || caller.organizationId !== session.organizationId) {
      throw new ForbiddenException();
    }
    return {
      id: caller.id,
      role: caller.role as Role,
      organizationId: caller.organizationId,
      accountId: caller.accountId,
    };
  }
}
